src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -119,72 +119,88 @@ public override Type GetConcreteType(Type parentType)
             object instance = collectionPropertyInfo.DeclaredTypeClassInfo.CreateObject();
 
-            if (instance is IList instanceOfIList)
+            try
             {
-                foreach (object item in sourceList)
+                if (instance is IList instanceOfIList && !instanceOfIList.IsReadOnly)
                 {
-                    instanceOfIList.Add(item);
+                    foreach (object item in sourceList)
+                    {
+                        instanceOfIList.Add(item);
+                    }
+                    return instanceOfIList;
                 }
-                return instanceOfIList;
-            }
-            else if (instance is ICollection<TRuntimeProperty> instanceOfICollection)
-            {
-                foreach (TRuntimeProperty item in sourceList)
+                else if (instance is ICollection<TRuntimeProperty> instanceOfICollection && !instanceOfICollection.IsReadOnly)
                 {
-                    instanceOfICollection.Add(item);
+                    foreach (TRuntimeProperty item in sourceList)
+                    {
+                        instanceOfICollection.Add(item);
+                    }
+                    return instanceOfICollection;
                 }
-                return instanceOfICollection;
-            }
-            else if (instance is Stack<TRuntimeProperty> instanceOfStack)
-            {
-                foreach (TRuntimeProperty item in sourceList)
+                else if (instance is Stack<TRuntimeProperty> instanceOfStack)
                 {
-                    instanceOfStack.Push(item);
+                    foreach (TRuntimeProperty item in sourceList)
+                    {
+                        instanceOfStack.Push(item);
+                    }
+                    return instanceOfStack;
                 }
-                return instanceOfStack;
-            }
-            else if (instance is Queue<TRuntimeProperty> instanceOfQueue)
-            {
-                foreach (TRuntimeProperty item in sourceList)
+                else if (instance is Queue<TRuntimeProperty> instanceOfQueue)
                 {
-                    instanceOfQueue.Enqueue(item);
+                    foreach (TRuntimeProperty item in sourceList)
+                    {
+                        instanceOfQueue.Enqueue(item);
+                    }
+                    return instanceOfQueue;
                 }
-                return instanceOfQueue;
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
             }
 
             // TODO: Use reflection to support types implementing Stack or Queue.
 
-            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionPropertyInfo.DeclaredPropertyType, jsonPath);
+            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionPropertyInfo.DeclaredPropertyType, jsonPath, exception);
```

Comment:
We should throw a `NotSupportedException` here using https://github.com/dotnet/corefx/blob/6f8983d64c67fa86ac6f0338cde08d82e51cc805/src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs#L21